TrackletAlgorithm/VMRouterCM.h

```diff
@@ -0,0 +1,434 @@
+					passRZCut = !(comparison_rz > comparison_value);
+					passRZSpecialCut = !(comparison_rz < comparison_valueLMR);
+				} else if (Layer == 2) {
+					constexpr float comparison_value = (Layer) ? VMROUTERCUTZL2 / kz_cm[Layer-1] : 0;
+					passRZCut = !(comparison_rz < comparison_value);
+				} else if (Layer == 3 || Layer == 5) {
+					constexpr float comparison_value = (Layer) ? VMROUTERCUTZL1L3L5 / kz_cm[Layer-1] : 0;
+					passRZCut = !(comparison_rz > comparison_value);
+					std::cout<< comparison_rz << " > " << comparison_value << std::endl;
+				} else if (Disk == 1 || Disk == 3) {
+					constexpr float comparison_value = VMROUTERCUTRD1D3 / kr;
+					constexpr float comparison_value2 = 2*N_DISK; // 2*int(N_DSS_MOD) in emulation
+					passRZCut = !(comparison_rz > comparison_value) && !(comparison_rz < comparison_value2);
+				}
+
+				// Write the stubs that pass the RZ cuts
+				if (passRZCut || passRZSpecialCut) {
+
+					AllStubInner<OutType> allstubinner;
+					allstubinner = (disk2S) ? static_cast<ap_uint<allstubinner.getWidth()>>(stubDisk2S.raw()) << (allstubinner.getWidth() - stubDisk2S.getWidth()) 
+								: static_cast<ap_uint<allstubinner.getWidth()>>(stub.raw()) << (allstubinner.getWidth() - stub.getWidth()); // shift by finephi and index
+
+					allstubinner.setIndex(i);
+					auto phicorr = getPhiCorr<InType>(stub.getPhi(), stub.getR(), stub.getBend(), phiCorrTable); // Corrected phi, i.e. phi at nominal radius (what about disks?)
+					allstubinner.setFinePhi(phicorr.range(phicorr.length() - 1, phicorr.length() - allstubinner.getFinePhi().length())); // top 8 bits of phicorr
+
+					constexpr unsigned int phicutmax = (Layer == 1) ? 4 : 6; // I have no idea what these numbers are, see emulation
+					constexpr unsigned int phicutmin = (Layer == 1) ? 4 : 2;
+
+					constexpr unsigned int nbitsall = (Layer) ? nbitsallstubs[Layer-1] : nbitsallstubs[N_LAYER+Disk-1]; // Number of bits for the number of Alltub memories in a layer/disk
+
+					auto iphipos = phicorr.range(phicorr.length() - nbitsall -1, phicorr.length() - (nbitsall + phiRegSize)); // Top three bits after the allstub bits 
+					unsigned int inner_mem_index = 0; // Keeps track of which allstub inner memory to write to
+
+
+					for (int n = 0; n < maskASIsize; n++) {
+#pragma HLS UNROLL
+
+						if (maskASI[n]) {
+							bool passPhiCut = false;
+							bool passSpecialCut = (Layer != 1) ? true : ((n > 5) && passRZSpecialCut) || (!(n > 5) && passRZCut); // For layer 1 the LMR memories have different cuts
+
+							if (n==6 || n==9) { // L
+								passPhiCut = !(iphipos>=phicutmin);
+							} else if (n == 8 || n==11) { // R
+								passPhiCut = !(iphipos<phicutmax);
+							} else if (n == 0 || n==3 || n==5) { // A, D, F
+								passPhiCut = !(iphipos<4);
+							} else if (n == 1 || n==2 || n==4) { // B, C, E
+								passPhiCut = !(iphipos>=4);
+							} else { // M
+								passPhiCut = true;
+							}
```

Comment:
There are several magic numbers here related to the index into `maskASI`. And based on your comments, I guess this index corresponds to the letters in the names of the AS memories? If that's the case, then would it make sense to define an `enum` somewhere so that you can use letters instead of magic numbers in these comparisons?